File: src/main/java/aurocosh/divinefavor/common/potions/blessings/PotionWarpingPresence.java
package aurocosh.divinefavor.common.potions.blessings;

import aurocosh.divinefavor.common.custom_data.player.PlayerData;
import aurocosh.divinefavor.common.custom_data.player.data.presence.warping.WarpingPresenceData;
import aurocosh.divinefavor.common.item.calling_stones.ModCallingStones;
import aurocosh.divinefavor.common.potions.base.potion.ModPotion;
import aurocosh.divinefavor.common.potions.common.ModBlessings;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraftforge.event.entity.living.EnderTeleportEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

@Mod.EventBusSubscriber
public class PotionWarpingPresence extends ModPotion {

    public PotionWarpingPresence() {
        super("warping_presence", true, 0x7FB8A4);
    }

    @Override
    protected void onPotionAdded(EntityLivingBase livingBase) {
        super.onPotionAdded(livingBase);
        if(!(livingBase instanceof EntityPlayer))
            return;
        EntityPlayer player = (EntityPlayer) livingBase;
        WarpingPresenceData presenceData = PlayerData.get(player).getWarpingPresenceData();
        presenceData.reset();
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public static void onEntityDamaged(EnderTeleportEvent event) {
        Entity entity = event.getEntity();
        if (!(entity instanceof EntityPlayer))
            return;
        EntityPlayer player = (EntityPlayer) entity;
        if(!player.isPotionActive(ModBlessings.warping_presence))
            return;
        WarpingPresenceData presenceData = PlayerData.get(player).getWarpingPresenceData();
        if(presenceData.tryLuck() && !player.world.isRemote){
            player.removePotionEffect(ModBlessings.warping_presence);
            player.addItemStackToInventory(new ItemStack(ModCallingStones.calling_stone_endererer));
        }
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/potions/presences/PotionWarpingPresence.kt
package aurocosh.divinefavor.common.potions.presences

import aurocosh.divinefavor.DivineFavor
import aurocosh.divinefavor.common.item.common.ModCallingStones
import aurocosh.divinefavor.common.lib.extensions.divinePlayerData
import aurocosh.divinefavor.common.potions.base.potion.ModPotion
import aurocosh.divinefavor.common.potions.common.ModBlessings
import aurocosh.divinefavor.common.spirit.ModSpirits
import aurocosh.divinefavor.common.util.UtilSpirit
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraftforge.event.entity.living.EnderTeleportEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

@Mod.EventBusSubscriber(modid = DivineFavor.MOD_ID)
class PotionWarpingPresence : ModPotion("warping_presence", 0x7FB8A4) {

    override fun onPotionAdded(livingBase: EntityLivingBase) {
        super.onPotionAdded(livingBase)
        if (livingBase is EntityPlayer)
            livingBase.divinePlayerData.warpingPresenceData.reset()
    }

    companion object {

        @SubscribeEvent(priority = EventPriority.LOWEST)
        @JvmStatic
        fun onEntityDamaged(event: EnderTeleportEvent) {
            val entity = event.entity as? EntityPlayer ?: return
            if (!entity.isPotionActive(ModBlessings.warping_presence))
                return
            if (!entity.world.isRemote && entity.divinePlayerData.warpingPresenceData.tryLuck()) {
                entity.removePotionEffect(ModBlessings.warping_presence)
                UtilSpirit.convertMarksToInvites(entity, ModSpirits.endererer, ModCallingStones.calling_stone_endererer)
                MaterialPresence.onInviteGiven(entity)
            }
        }
    }
}
